### ISSUE TITLE
Profile a validation cycle: nothing on that path is compiled, including the generator

### ISSUE THREAD

**Issue by YeapJieShen** [YeapJieShen/single-image-super-resolution#226]
## The suspicion

`torch.compile` may never reach the VGG feature extractor, and validation may be paying for it.

## What the code actually does

Compilation wraps **one module and one module only**:

```python
compiled = torch.compile(self.model, backend=backend, mode=...) if backend else None
```

`self.model` is the SR generator. Never compiled, therefore:

- the **criterion** — including `VGG19FeatureLoss`, which is the content loss for every VGG
  recipe (SRGAN, and SRResNet-VGG22)
- the **discriminator**
- the **perceptual metrics**, LPIPS and DISTS, which are themselves AlexNet/VGG16 forwards

And dispatch is gated on training mode:

```python
model_fn = self._compiled if self.training and self._compiled is not None else self.model
```

So at validation **even the SR model runs eager**. The whole validation path is uncompiled.

The VGG forward is genuinely on the validation path, not just the training one: `validation_step`
calls `_step`, and `_step` calls `self.criterion(sr_model_out, hr_for_loss)` unconditionally.

## Why this is worth measuring rather than assuming

Validation already dominates wall-clock on an adversarial run, and it runs on **full-size
images**, not 96x96 training crops — so the VGG forward there is far larger per call than the one
inside a training step.

But the gate on `self.training` is not an oversight, and any fix has to answer it: training uses
one fixed patch shape, while validation and benchmark loaders see widely varying image sizes.
A naively compiled VGG would recompile per unique shape and could easily cost more than it saves.
`dynamic=True`, or bucketing, is the part that needs evidence.

## Scope

Profile first, compile second. Filing the profiling as the deliverable, because it is what
decides whether the compilation is worth doing at all:

1. **Where does a validation cycle's time actually go** — SR forward, VGG content loss, LPIPS,
   DISTS, metric computation, image logging. A breakdown, not a total.
2. **How many distinct input shapes** a validation cycle presents, which is what sets the
   recompilation bill.
3. Only then: whether compiling the criterion and/or the perceptual backbones pays, and under
   which `dynamic` setting.

## Acceptance

- [ ] A per-component breakdown of one validation cycle, on real data
- [ ] The distinct-shape count for the validation and benchmark loaders
- [ ] A before/after time-per-cycle figure for whatever change follows, or a recorded decision
      that compilation does not pay and why

## Measurement bar

Two standing traps apply and both have bitten this project before:

- A throughput number needs its provenance: which data, which power state, warmup excluded, and
  **medians rather than means** — a mean hides one-time costs like autotune and worker spawn.
- **A polled rate is quantised by `log_every_n_steps`** and the quantisation forges precision:
  repeated windows agree to five significant figures while resolving nothing. Divide the step
  delta by the logging cadence and report that count.

Blocked on the GPU until the running acceptance job finishes.

**Comment by YeapJieShen** [comment]
**Split.** The training-path half is now #227.

This issue keeps the open-ended question: **where does a validation cycle's time actually go, and is any of it compilable given that validation sees many image sizes?** That is an investigation which may well conclude "not worth it".

#227 keeps the part with no such objection: the training path has exactly one crop shape, so compiling the content loss there faces no recompilation bill. It is a clean candidate and should not wait on this.

Both are fed by the same profiling pass, so the instrumentation is written once.

**Comment by YeapJieShen** [comment]
## Profiled. The suspicion was right about the gate and wrong about where the time is.

One full adversarial validation cycle — 219 images across the primary split and
all three benchmark sets, real data, on mains, GPU otherwise idle. Hand-placed
CUDA-synced timers rather than `torch.profiler`, whose own overhead lands
hardest on exactly the launch-bound regime this box is sensitive to. Warmup
cycle discarded; medians per image.

### Where a validation cycle actually goes

| component | total | share | median / image |
|---|---:|---:|---:|
| **DISTS** | 165.22 s | **47.6%** | 37.44 ms |
| **VGG content loss** | 140.23 s | **40.4%** | 31.69 ms |
| SR forward | 24.55 s | 7.1% | 26.56 ms |
| PSNR + SSIM | 9.73 s | 2.8% | 11.50 ms |
| host-to-device | 4.00 s | 1.2% | 3.45 ms |
| LPIPS | 3.27 s | 0.9% | 9.53 ms |
| **total (timed)** | **347.00 s** | | |

**DISTS and the VGG content loss are 88% of the cycle.** The SR generator — the
one module `torch.compile` currently wraps, and the one the `self.training` gate
excludes here — is **7.1%**.

**DISTS costs 3.9x LPIPS per image** (37.44 vs 9.53 ms) despite being the same
kind of metric on the same images. It is the single most expensive thing on the
validation path, and that was not predicted anywhere.

### The first cycle costs 2.9x the steady-state one

Warmup cycle **1030.4 s**; the next **350.3 s**. That gap is cuDNN autotuning a
conv algorithm per distinct input shape, plus lazy init. A run's *first*
validation is ~17 minutes and every later one ~5.8 — so a mean over cycles
misreports both.

### The recompilation bill

| loader | images | distinct LR shapes |
|---|---:|---:|
| val | 100 | **24** |
| Set5 | 5 | 5 |
| Set14 | 14 | 10 |
| BSD100 | 100 | 2 |

**41 distinct shapes per cycle.**

## Decision: do not compile the validation path

Recorded as the issue asked for, rather than left implicit.

- **Compiling the generator here cannot pay.** It is 7.1% of the cycle. Even a
  2x speedup on it returns 3.5%, against 41 shape specialisations. The
  `self.training` gate is doing the right thing and should stay.
- **The 88% is the wrong shape to compile.** DISTS and the VGG stack are where
  the time is, but they are also what sees all 41 shapes. `dynamic=True` trades
  away most of the gain; per-shape specialisation pays compilation 41 times per
  process for modules that are already dense convolution.
- **The cheap lever is not compilation at all.** DISTS at 47.6% is a scope
  question — which images need it — not a codegen one. Filing that separately
  rather than smuggling it in here.

**Where compilation can still pay is the training path**, and that is #227's
scope: exactly one input shape, so no recompilation bill, and the same VGG stack
that costs 40% of a validation cycle also runs twice per training step.
